### PR TITLE
more consistent errors

### PR DESCRIPTION
**Pull request by bnjjj**
related to #2101 
This is a draft/attempt to have more consistent errors in the router following this spec https://www.apollographql.com/docs/apollo-server/data/errors/ with the error extension code.

### CODE REVIEW

File: apollo-router/src/error.rs [hunk 1]
```diff
@@ -132,6 +144,26 @@ impl FetchError {
...
+                "SubrequestUnexpectedPatchResponse"
+            }
+            FetchError::SubrequestHttpError { .. } => "SubrequestHttpError",
+            FetchError::ExecutionFieldNotFound { .. } => "ExecutionFieldNotFound",
+            FetchError::ExecutionPathNotFound { .. } => "ExecutionPathNotFound",
+            FetchError::CompressionError { .. } => "CompressionError",
+            #[cfg(test)]
+            FetchError::ExecutionInvalidContent { .. } => "ExecutionInvalidContent",
+        }
+        .to_shouty_snake_case()
```
Comment (BrynCooke) [inline apollo-router/src/error.rs:163]: Can we not use this and instead just make the string constants shouty snake case? That way if we have to search for these constants we'll find them. Same goes for all other places where this is used.

**Review by garypen — COMMENTED** [review]
I like the look of this PR. I'd like to see some examples of how errors will look before/after. I can see that a bit from the way tests have changed, but it would be nice to see it written out maybe in an issue or at least on the PR.

If we made this change would it effectively be a router 2.0 change? I mean do we think this is part of our public API?

**Comment by bnjjj** [comment]
@garypen I don't think it should be a 2.0 change. Because we didn't really remove something, we just added more context to errors so it's additive

File: apollo-router/src/plugin/test/mock/subgraph.rs [hunk 2]
```diff
@@ -101,6 +101,7 @@ impl Service<SubgraphRequest> for MockSubgraph {
                     "couldn't find mock for query {}",
                     serde_json::to_string(&req.subgraph_request.body()).unwrap()
                 ))
+                .extension_code("FETCH_ERROR".to_string())
```
Comment (Geal) [inline apollo-router/src/plugin/test/mock/subgraph.rs:104]: maybe make the common extension code constants?

**Review by Geal — APPROVED** [review]

File: apollo-router/src/plugin/test/mock/subgraph.rs (hunk 2 above)
Comment (bnjjj) [reply apollo-router/src/plugin/test/mock/subgraph.rs:104]: I tempted to say we don't care because the only places we use it is in tests and mock

**Review by BrynCooke — APPROVED** [review]

File: apollo-router/src/graphql.rs [hunk 3]
```diff
@@ -93,13 +95,17 @@ impl Error {
     /// * `.build()`
     ///   Finishes the builder and returns a GraphQL [`Error`].
     #[builder(visibility = "pub")]
-    fn new(
+    fn new<T: Into<String>>(
         message: String,
         locations: Vec<Location>,
         path: Option<Path>,
+        extension_code: T,
```
Comment (yanns) [inline apollo-router/src/graphql.rs:102]: This new method should be added in the documentation:
```
    /// Returns a builder that builds a GraphQL [`Error`] from its components.
    ///
    /// Builder methods:
    ///
    /// * `.message(impl Into<`[`String`]`>)`
    ///   Required.
    ///   Sets [`Error::message`].
    ///
    /// * `.locations(impl Into<`[`Vec`]`<`[`Location`]`>>)`
    ///   Optional.
    ///   Sets the entire `Vec` of [`Error::locations`], which defaults to the empty.
    ///
    /// * `.location(impl Into<`[`Location`]`>)`
    ///   Optional, may be called multiple times.
    ///   Adds one item at the end of [`Error::locations`].
    ///
    /// * `.path(impl Into<`[`Path`]`>)`
    ///   Optional.
    ///   Sets [`Error::path`].
    ///
    /// * `.extensions(impl Into<`[`serde_json_bytes::Map`]`<`[`ByteString`]`, `[`Value`]`>>)`
    ///   Optional.
    ///   Sets the entire [`Error::extensions`] map, which defaults to empty.
    ///
    /// * `.extension(impl Into<`[`ByteString`]`>, impl Into<`[`Value`]`>)`
    ///   Optional, may be called multiple times.
    ///   Adds one item to the [`Error::extensions`] map.
    ///
    /// * `.build()`
    ///   Finishes the builder and returns a GraphQL [`Error`].
```
We need to add `.extension_code(impl Into<`[`String`]`>)`

Comment (bnjjj) [reply apollo-router/src/graphql.rs:102]: I created an issue #2469 Thanks a lot

**Pull request merged**